stream_chat/types/stream_response.py

```diff
@@ -1,20 +1,76 @@
-from typing import Dict
+from datetime import datetime, timezone
+from typing import Any, Dict, Optional
 
+from stream_chat.types.rate_limit import RateLimitInfo
 
-class StreamResponse(Dict):
-    pass
 
+class StreamResponse(dict):
+    """
+    A custom dictionary where you can access the response data the same way
+    as a normal dictionary. Additionally, we expose some methods to access other metadata.
 
-"""
-Ideally, this would look something like this:
+    ::
 
-class StreamResponse(TypedDict, allow_extra=True):
-    duration: str
+        resp = client.get_app_settings()
+        print(resp["app"]["webhook_url"])
+        # "https://mycompany.com/webhook"
+        rate_limit = resp.rate_limit()
+        print(rate_limit.remaining)
+        # 99
+        headers = resp.headers()
+        print(headers["content-type"])
+        # "application/json;charset=utf-8"
+        status_code = resp.status_code()
+        print(status_code)
+        # 200
+    """
 
-The problem is that Python is not as smart as TypeScript today. `allow_extra` flag
-doesn't exist, only `total=False` which is a bit different.
+    def __init__(
+        self, response_dict: Dict[str, Any], headers: Dict[str, Any], status_code: int
+    ) -> None:
+        self.__response_dict = response_dict
+        self.__headers = headers
+        self.__status_code = status_code
+        self.__rate_limit: Optional[RateLimitInfo] = None
+        limit, remaining, reset = (
+            headers.get("x-ratelimit-limit"),
+            headers.get("x-ratelimit-remaining"),
+            headers.get("x-ratelimit-reset"),
+        )
+        if limit and remaining and reset:
+            self.__rate_limit = RateLimitInfo(
+                limit=int(limit),
+                remaining=int(remaining),
+                reset=datetime.fromtimestamp(float(reset), timezone.utc),
+            )
 
-We'll just inherit from Dict until they solve this problem.
+        super(StreamResponse, self).__init__(self.__response_dict)
 
-https://github.com/python/mypy/issues/4617
-"""
+    def rate_limit(self) -> Optional[RateLimitInfo]:
+        """Returns the ratelimit info of your API operation."""
+        return self.__rate_limit
+
+    def headers(self) -> Dict[str, Any]:
+        """Returns the headers of the response."""
+        return self.__headers
+
+    def status_code(self) -> int:
+        """Returns the HTTP status code of the response."""
+        return self.__status_code
+
+    def response_body(self) -> Dict[str, Any]:
+        """Returns the response body as a dictionary."""
+        return self.__response_dict
+
+    def __str__(self) -> str:
+        return str(self.__response_dict)
```

Comment:
Since you're calling the `super(StreamResponse, self).__init__(self.__response_dict)`, I don't really see why you would even need this `self.__response_dict` attribute anymore. the `StreamResponse` *is* the response_dict now, right?

Reply:
yup, good point.  updated 👍